tests/test_mpls.py

```diff
@@ -0,0 +1,591 @@
+
+        # remove route entry
+        if self.mpls_appdb_mode():
+            self.remove_route_entry(prefix)
+        else:
+            dvs.runcmd("ip route del 2.2.2.0/24 nexthop encap mpls 200 via inet 10.0.0.1 dev Ethernet0 nexthop encap mpls 201 via inet 10.0.0.5 dev Ethernet8")
+
+        # check application database
+        self.pdb.wait_for_deleted_entry("ROUTE_TABLE", prefix)
+        time.sleep(1)
+
+        # check ASIC route database
+        self.check_route_entries(False, [prefix])
+        self.check_nexthop_group(False, nhg)
+        self.check_nexthop(False, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_PUSH", "1:200")
+        self.check_nexthop(False, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.5", "SAI_OUTSEG_TYPE_PUSH", "1:201")
+
+        self.teardown_mpls(dvs)
+
+    def test_RouteAddRemoveMplsRouteSwapNHG(self, dvs, testlog):
+        self.setup_mpls(dvs, True)
+
+        # add route entry
+        label = "200"
+        if self.mpls_appdb_mode():
+            fieldValues = {"nexthop": "10.0.0.1,10.0.0.5", "ifname": "Ethernet0,Ethernet8", "mpls_nh": "swap+201,swap+202", "mpls_pop": "1"}
+            self.create_inseg_entry(label, fieldValues)
+        else:
+            dvs.runcmd("ip -f mpls route add 200 nexthop as 201 via inet 10.0.0.1 dev Ethernet0 nexthop as 202 via inet 10.0.0.5 dev Ethernet8")
+
+        # check application database
+        self.pdb.wait_for_entry("LABEL_ROUTE_TABLE", label)
+        time.sleep(1)
+
+        # check ASIC inseg database
+        self.check_inseg_entries(True, [label])
+        nhg = self.check_inseg_nexthop_group(label, 2)
+        self.check_nexthop(True, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_SWAP", "1:201")
+        self.check_nexthop(True, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.5", "SAI_OUTSEG_TYPE_SWAP", "1:202")
+
+        # remove route entry
+        if self.mpls_appdb_mode():
+            self.remove_inseg_entry(label)
+        else:
+            dvs.runcmd("ip -f mpls route del 200 nexthop as 201 via inet 10.0.0.1 dev Ethernet0 nexthop as 202 via inet 10.0.0.5 dev Ethernet8")
+
+        # check application database
+        self.pdb.wait_for_deleted_entry("LABEL_ROUTE_TABLE", label)
+        time.sleep(1)
+
+        # check ASIC inseg database
+        self.check_inseg_entries(False, [label])
+        self.check_nexthop_group(False, nhg)
+        self.check_nexthop(False, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_SWAP", "1:201")
+        self.check_nexthop(False, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.5", "SAI_OUTSEG_TYPE_SWAP", "1:202")
+
+        self.teardown_mpls(dvs)
+
+    def test_RouteAddRemoveMplsRoutePopNHG(self, dvs, testlog):
+        self.setup_mpls(dvs, True)
+
+        # add route entry
+        label = "200"
+        if self.mpls_appdb_mode():
+            fieldValues = {"nexthop": "10.0.0.1,10.0.0.5", "ifname": "Ethernet0,Ethernet8", "mpls_nh": "swap+,swap+", "mpls_pop": "1"}
```

Comment:
same here, "swap+", is it required?

Reply:
@smahesh Yes, as previously states, you're the one who stated that POP was programmed in SAI as swap action with empty labelstack.

Reply:
@smaheshm Programming is updated per latest discussion.